### PR TITLE
Electricity Maps Data Source Implementation

### LINKED ISSUE #71: [Feature Contribution]: electricityMap Plugin
### What happened?

This feature contribution will add electricityMap plugin support and various improvements to the Carbon Aware SDK WebAPI and CLI.  This plugin should be similar to Watttime plugin.
Electricity map supports more countries' data including many Asia such as India, China and others which Watttime doesn't support.  (#67 for details)
As discussed in a weekly meeting (5th July), target to merge into v1.1.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Feature Commitment

- [X] I commit to contributing this feature as a PR and working with the GSF to merge this feature into the Carbon Aware SDK.

### PR DESCRIPTION
**Pull request by JenMadiedo**
Issue Number: #71 

## Summary
Implementation of the ElectricityMaps Client

## Changes

- Add ElectricityMaps Client + Data source implementation for `forecast` routes
- Add ElectricityMaps Client mocker
- Add unit tests for client and data source
- Update WebAPI and CLI controller integration tests for ElectricityMaps
- Update docs

## Checklist

- [X] Local Tests Passing?
- [X] CICD and Pipeline Tests Passing?
- [X] Added any new Tests?
- [X] Documentation Updates Made (Will be made with data source PR)?
- [ ] Are there any API Changes? If yes, please describe below.
- [ ] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?
If yes, what are the expected API Changes? Please link to an API-Comparison workflow with the API Diff.

## Is this a breaking change?
If yes, what workflow does this break? 

## Anything else?
Other comments, collaborators, etc.
> Please follow [GitHub's suggested syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to link Pull Requests to Issues via keywords

This PR Closes #71

### CODE REVIEW

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+        Assert.That(dataPoint?.CarbonIntensity.ToString(), Is.EqualTo("999"));
+        Assert.That(dataPoint?.EmissionFactorType, Is.EqualTo(EmissionsFactor.Lifecycle));
+        Assert.IsFalse(dataPoint?.IsEstimated);
+        Assert.IsEmpty(dataPoint?.EstimationMethod);
+        Assert.That(new DateTimeOffset(2099, 1, 1, 0, 0, 0, TimeSpan.Zero), Is.EqualTo(dataPoint?.CreatedAt));
+    }
+
+    [TestCase("x-token-header", "faketoken", TestName = "Client config test, token header + token value valid")]
+    [TestCase("", "", TestName = "Client config test, Empty token header + empty token value valid")]
+    public void ClientTest_With_ServiceCollection_VariedConfiguration_DoesNotThrow(string tokenHeader, string tokenValue)
```
Comment (bderusha) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:312]: These `ServiceCollection` tests feel out of place.  Should they be in their own file?

It's not quite clear what the intent of the test are. It feels like they should maybe be direct `ElectricityMapsClientConfiguration_Validate` tests and there should be 1 ServiceCollection test that confirms that `Validate` is called, rather than doing all of this setup.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+    private void CreateHttpClient(Func<HttpRequestMessage, Task<HttpResponseMessage>> requestDelegate)
+    {
+        this.MessageHandler = new MockHttpMessageHandler(requestDelegate);
+        this.HttpClient = new HttpClient(this.MessageHandler);
+        this.HttpClientFactory = Mock.Of<IHttpClientFactory>();
+        Mock.Get(this.HttpClientFactory).Setup(h => h.CreateClient(IElectricityMapsClient.NamedClient)).Returns(this.HttpClient);
+    }
+
+    private class MockHttpMessageHandler : HttpMessageHandler
```
Comment (bderusha) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:443]: Consider using `Moq.Protected` to manage the setup for `SendAsync` rather than writing our own class

- [Moq/moq4#wiki](https://github.com/Moq/moq4/wiki/Quickstart#miscellaneous:~:text=Moq%204.8%20and%20later%20allows%20you%20to%20set%20up%20protected%20members)
- https://kagamine.dev/en/mock-httpclient-the-easy-way/

File: src/CarbonAware.LocationSources/src/Model/NamedGeoposition.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+using CarbonAware.LocationSources.Exceptions;
+using CarbonAware.Model;
+using System.Globalization;
+
+namespace CarbonAware.LocationSources.Model;
+
+public class NamedGeoposition
```
Comment (bderusha) [inline src/CarbonAware.LocationSources/src/Model/NamedGeoposition.cs:7]: This class needs unit tests for `AssertValid` and `explicit operator Location`

File: src/CarbonAware.WebApi/test/integrationTests/CarbonAware.WebApi.IntegrationTests.csproj [hunk 4]
```diff
@@ -26,7 +26,8 @@
   <ItemGroup>
     <ProjectReference Include="..\..\..\CarbonAware.DataSources\CarbonAware.DataSources.Json\mock\CarbonAware.DataSources.Json.Mocks.csproj" />
     <ProjectReference Include="..\..\..\CarbonAware.DataSources\CarbonAware.DataSources.WattTime\mock\CarbonAware.DataSources.WattTime.Mocks.csproj" />
-    <ProjectReference Include="..\..\src\CarbonAware.WebApi.csproj" />
+	<ProjectReference Include="..\..\..\CarbonAware.DataSources\CarbonAware.DataSources.ElectricityMaps\mock\CarbonAware.DataSources.ElectricityMaps.Mocks.csproj" />
```
Comment (bderusha) [inline src/CarbonAware.WebApi/test/integrationTests/CarbonAware.WebApi.IntegrationTests.csproj:29]: NIT: indentation

File: src/CarbonAware.WebApi/test/integrationTests/CarbonAwareControllerTests.cs [hunk 5]
```diff
@@ -49,8 +48,10 @@ public async Task FakeEndPoint_ReturnsNotFound()
     [TestCase("2021-12-25", "2021-12-26", "westus")]
     public async Task BestLocations_ReturnsOK(DateTimeOffset start, DateTimeOffset end, string location)
     {
+        IgnoreTestForDataSource("data source does not implement '/emissions/forecasts/current'", DataSourceType.ElectricityMaps);
```
Comment (bderusha) [inline src/CarbonAware.WebApi/test/integrationTests/CarbonAwareControllerTests.cs:51]: Update the message to reflect the test, here and throughout this file.

File: src/CarbonAware/src/Configuration/ConfigurationExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+using Microsoft.Extensions.Configuration;
+
+namespace CarbonAware.Configuration;
+
+public static class ConfigurationExtensions
```
Comment (bderusha) [inline src/CarbonAware/src/Configuration/ConfigurationExtensions.cs:5]:
```suggestion
public static class DataSourcesConfigurationExtensions
```

**Review by bderusha — CHANGES_REQUESTED** [review]
Big PR. Big effort. Great work on this.

Lots of little comments to manage. Let's coordinate on how we want to handle owning the updates

File: docs/configuration.md [hunk 7]
```diff
@@ -160,6 +166,34 @@ info: CarbonAware.DataSources.Json.JsonDataSource[0]
     Reading Json data from /app/data-sources/json/mycustomfile.json
 ```
 
+### ElectricityMaps Configuration
+
+If using the ElectricityMaps data source, ElectricityMaps configuration is required.
+
+```json
```
Comment (JenMadiedo) [inline docs/configuration.md:173]: Here is a spot we may want to note the 2 different configurations (for token users vs. free trial users)

File: docs/configuration.md [hunk 8]
```diff
@@ -160,6 +166,34 @@ info: CarbonAware.DataSources.Json.JsonDataSource[0]
...
+
+#### API Token
+
+The Electricity Maps token you receive with your account
+
+#### baseUrl
+
+The url to use when connecting to Electricity.  Defaults to ["https://api.electricitymap.org/v3/"]("https://api.electricitymap.org/v3/").
+
+In normal use, you shouldn't need to set this value, but this value can be used to enable integration testing scenarios or if the ElectricityMaps url should change in the future.
```
Comment (JenMadiedo) [inline docs/configuration.md:195]: Add note here that "normal use" is with a token, will need to be updated for free trial use

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/mock/ElectricityMapDataSourceMocker.cs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                Response.Create()
+                    .WithStatusCode(HttpStatusCode.OK)
+                    .WithHeader("Content-Type", MediaTypeNames.Application.Json)
+                    .WithBody(jsonBody)
+        );
+    }
+
+    private string GetZoneGivenLatLong(decimal latitutde, decimal longitude)
+    {
+        // TODO: Update with lookup code.
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/mock/ElectricityMapDataSourceMocker.cs:135]: Agree let's remove. I think this was back when we thought we'd have a manual mapping for ^^, given that we won't makes sense to nix.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/ElectricityMapsClient.cs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            _log.LogDebug("Built url {result} from url {url} and query string parameters {parameters}", result, url, string.Join(";", queryStringParams.Select(k => $"\"{k.Key}\":\"{k.Value}\"")));
+        }
+        return result;
+    }
+
+    // Checks the current supported client's endpoint paths.
+    private async Task CheckZonesAllowedForPath(string path, Dictionary<string, string> parameters)
+    {
+        // Trial user won't have access to zones endpoint, exit
+        if (this._configuration.BaseUrl.Contains(BaseUrls.TrialBaseUrl)) return;
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/ElectricityMapsClient.cs:163]: The goal is to avoid the performance hit of trying to access ElectricityMaps when we know for a fact it will fail. Instead, just exit the check all together, as the user doesn't need this check, it's more to give them better error messages if the subsequent call fails

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/IElectricityMapsClient.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    public const string NamedClient = "ElectricityMapsClient";
+
+    /// <summary>
+    /// Async method to get the most recent 24 hour forecasted emission data for a given latitude and longitude.
+    /// </summary>
+    /// <param name="latitude">Latitude for query</param>
+    /// <param name="longitude">Longitude for query</param>
+    /// <returns>A <see cref="Task{ForecastedCarbonIntensityData}"/> which contains forecasted emissions data points.</returns>
+    /// <exception cref="ElectricityMapsClientException">Can be thrown when errors occur connecting to ElectricityMaps client.  See the ElectricityMapsClientException class for documentation of expected status codes.</exception>
+    public Task<ForecastedCarbonIntensityData> GetCurrentForecastAsync(string latitude, string longitude);
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/IElectricityMapsClient.cs:19]: We could do that, it was more for this to match to similar functions in WattTime Client so that the mental model of what the function is doing can map between clients. Don't feel super strongly, more just giving context of what the thinking was behind that naming scheme.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/IElectricityMapsClient.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    public Task<ForecastedCarbonIntensityData> GetCurrentForecastAsync(string zoneName);
+
+    /// <summary>
+    /// Async method to get the most recent 24 hour observed emission data for a given latitude and longitude.
+    /// </summary>
+    /// <param name="latitude">Latitude for query</param>
+    /// <param name="longitude">Longitude for query</param>
+    /// <returns>A <see cref="Task{HistoryCarbonIntensityData}"/> which contains all emissions data points in the 24 hour period.</returns>
+    /// <exception cref="ElectricityMapsClientException">Can be thrown when errors occur connecting to ElectricityMaps client.  See the ElectricityMapsClientException class for documentation of expected status codes.</exception>
+    public Task<HistoryCarbonIntensityData> GetHistoryCarbonIntensityDataAsync(string latitude, string longitude);
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/IElectricityMapsClient.cs:36]: See response above, same reasoning: https://github.com/Green-Software-Foundation/carbon-aware-sdk/pull/194/files/d6ced7c2fc33004212696f0fed73b26e444b829b..ca621f962077ee6057477daec9172da20023e817#r1030602604

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Configuration/ElectricityMapsClientConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        if (!Uri.IsWellFormedUriString(this.BaseUrl, UriKind.Absolute))
+        {
+            throw new ConfigurationException($"{nameof(this.BaseUrl)} is not a valid absolute url.");
+        }
+
+        var isTokenHeaderSet = !string.IsNullOrWhiteSpace(this.APITokenHeader);
+        var isTokenSet = !string.IsNullOrWhiteSpace(this.APIToken);
+
+        // Either provide full auth config or no auth config, not a combo of half and half
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Configuration/ElectricityMapsClientConfiguration.cs:39]: ElectricityMaps lets you make calls without authentication. The zones endpoint for example I believe lets you do it. So, it was more to match that behavior, even though the calls within the library do require tokens. So depends if we want to enable that zones call in the future or not

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/ElectricityMapsDataSource.cs [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            var emissionsForecast = (EmissionsForecast) forecast;
+            var duration = emissionsForecast.GetDurationBetweenForecastDataPoints();
+            emissionsForecast.Location = location;
+            emissionsForecast.RequestedAt = requestedAt;
+            emissionsForecast.ForecastData = emissionsForecast.ForecastData.Select(d => 
+            {
+                d.Location = location.Name;
+                d.Duration = duration;
+                return d;
+            });
```
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/ElectricityMapsDataSource.cs:71]: All this (lines 61-71) feel like they should live in the `EmissionsForcast` class because it's all about how to build that EmissionsForecast object. Maybe have a function called `ToEmissionsForecast(ForecastedCarbonIntensityData forecast, Location location)` that does everything those 10 lines do internally. Because things like calculating the duration from the forecast and setting `RequestedAt` to the default don't feel like they're the datasources' job.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/EmissionsFactor.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+using System.Text.Json.Serialization;
+
+namespace CarbonAware.DataSources.ElectricityMaps.Model;
+
+[JsonConverter(typeof(JsonStringEnumConverter))]
+public enum EmissionsFactor
```
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/EmissionsFactor.cs:6]: 🐑 include link to `ElectricityMaps` webpage explaining what these are, not intuitive unless you've seen it in their docs

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/HistoryCarbonIntensityData.cs [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    /// Estimated data.
+    /// </summary>
+    [JsonPropertyName("isEstimated")]
+    public Boolean IsEstimated { get; init; }
+
+    /// <summary>
+    /// Estimated method.
+    /// </summary>
+    [JsonPropertyName("estimationMethod")]
+    public string? EstimationMethod { get; init; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/HistoryCarbonIntensityData.cs:70]: I think we actually should not set the default, think we meant to leave one or the other, not both

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 17]
```diff
@@ -0,0 +1,457 @@
+using CarbonAware.Configuration;
+using CarbonAware.DataSources.ElectricityMaps.Client;
+using CarbonAware.DataSources.ElectricityMaps.Configuration;
+using CarbonAware.DataSources.ElectricityMaps.Constants;
+using CarbonAware.DataSources.ElectricityMaps.Model;
+using CarbonAware.Exceptions;
+using Castle.Core.Internal;
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:7]: 😕

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 18]
```diff
@@ -0,0 +1,457 @@
...
+        noAuthConfig_Trial.Setup(o => o.CurrentValue).Returns(() => new ElectricityMapsClientConfiguration() { BaseUrl = BaseUrls.TrialBaseUrl});
+        var trialClient = new ElectricityMapsClient(this.HttpClientFactory, noAuthConfig_Trial.Object, this.Log.Object);
+
+        // Act & Assert
+        Assert.ThrowsAsync<ElectricityMapsClientHttpException>(async () => await tokenClient.GetCurrentForecastAsync(TestZone));
+        Assert.ThrowsAsync<ElectricityMapsClientHttpException>(async () => await tokenClient.GetCurrentForecastAsync(TestLatitude, TestLongitude));
+        Assert.ThrowsAsync<ElectricityMapsClientHttpException>(async () => await tokenClient.GetHistoryCarbonIntensityDataAsync(TestZone));
+        Assert.ThrowsAsync<ElectricityMapsClientHttpException>(async () => await tokenClient.GetHistoryCarbonIntensityDataAsync(TestLatitude, TestLongitude));
+
+        Assert.ThrowsAsync<ElectricityMapsClientHttpException>(async () => await trialClient.GetCurrentForecastAsync(TestZone));
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:84]: They could, just figured they'd be together because all methods throw with not auth, independent of the trial vs. token status

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs (hunk 2 above)
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:443]: This is to mirror the WattTimeClient, would it be worth doing that as a separate cleanup on both later?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/ElectricityMapsDataSourceTests.cs [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    private DateTimeOffset _defaultDataStartTime { get; set; }
+
+    [SetUp]
+    public void Setup()
+    {
+        _logger = new Mock<ILogger<ElectricityMapsDataSource>>();
+        _electricityMapsClient = new Mock<IElectricityMapsClient>();
+        _locationSource = new Mock<ILocationSource>();
+        _dataSource = new ElectricityMapsDataSource(_logger.Object, _electricityMapsClient.Object, _locationSource.Object);
+        _defaultLocation = new Location() { Name = "eastus", Latitude = 34.123m, Longitude = 123.456m };
```
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/ElectricityMapsDataSourceTests.cs:34]: Why set this before each test instead of just once when the field is initialized?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.Registration/Configuration/ServiceCollectionExtensions.cs [hunk 20]
```diff
@@ -12,9 +13,7 @@ public static class ServiceCollectionExtensions
 {
     public static IServiceCollection AddDataSourceService(this IServiceCollection services, IConfiguration configuration)
     {
-        var dataSources = configuration.GetSection(DataSourcesConfiguration.Key).Get<DataSourcesConfiguration>() ?? new DataSourcesConfiguration();
```
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.Registration/Configuration/ServiceCollectionExtensions.cs]: What was the reason behind this change? Feels a bit out of place with this PR, was it an issue of not pulling in the latest datasource config changes?

File: src/CarbonAware.LocationSources/src/LocationSource.cs [hunk 21]
```diff
@@ -2,6 +2,7 @@
 using CarbonAware.Interfaces;
 using CarbonAware.Model;
 using CarbonAware.LocationSources.Configuration;
+using CarbonAware.LocationSources.Model;
```
Comment (JenMadiedo) [inline src/CarbonAware.LocationSources/src/LocationSource.cs:5]: May be worth including a comment in the general PR explaining why we're changing the `LocationSources` project and what changes we had to make

File: src/data/location-sources/custom-azure-zones.json [hunk 22]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (JenMadiedo) [inline src/data/location-sources/custom-azure-zones.json:1]: Should we add docs around this file somewhere?

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/ElectricityMapsDataSource.cs (hunk 14 above)
Comment (pritipath) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/ElectricityMapsDataSource.cs:71]: I dont think the CarbonAware model `EmissionsForecast` should know about the DataSource model classes. But I agree, we can create a method in the datasource class to do the conversion for a more readable code. Also, the duration calculation is done in `ForecastedCarbonIntensityData` not the datasource.

**Comment by YaSuenag** [comment]
Does this PR relate to #75 ?

I've not yet deep dive to this PR, but I concerned this PR deeply depends on commercial license of electricityMap.

I know all of features in the SDK needs commercial features in electricityMap, but I think it would be better we can use personal electricityMap license even if the SDK features are limited. #75 can be used with personal license. I think it is a big hurdle to get commercial license like WattTime plugin for personal users who want to use the SDK casually.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/HistoryCarbonIntensityData.cs [hunk 23]
```diff
@@ -0,0 +1,72 @@
+using System.Text.Json.Serialization;
+
+namespace CarbonAware.DataSources.ElectricityMaps.Model;
+
+/// <summary>
+/// History Carbon Intensity collection data.
+/// </summary>
+[Serializable]
+public record HistoryCarbonIntensityData
```
Comment (pritipath) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/HistoryCarbonIntensityData.cs:9]: I am wondering if we should call this `RecentCarbonIntensityHistoryData` and the past range data (not part of this PR) as `PastCarbonIntensityHistoryData` to exactly mimic the Electricity Maps API spec. This way we have a clear distinction between the 2 models and is easy to decipher what each means.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/CarbonAware.DataSources.ElectricityMaps.Tests.csproj [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="coverlet.collector" Version="3.1.2">
+			<PrivateAssets>all</PrivateAssets>
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/CarbonAware.DataSources.ElectricityMaps.Tests.csproj:12]: Indentation looks fine on IDE, may be a github thing

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs (hunk 2 above)
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:443]: @bderusha  will circle back

File: src/data/location-sources/custom-azure-zones.json (hunk 22 above)
Comment (JenMadiedo) [reply src/data/location-sources/custom-azure-zones.json:1]: Will make this a template file so that this won't be picked up by config
@juzuluag 
- Add a story for overriding and how to manage multiple files etc...
- Make a template
- Update Docs

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/CarbonAware.DataSources.ElectricityMaps.Tests.csproj (hunk 24 above)
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/CarbonAware.DataSources.ElectricityMaps.Tests.csproj]: @pritipath will do this

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/ElectricityMapsClient.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/ElectricityMapsClient.cs:163]: @JenMadiedo will workaround the 500 failure post convo with bill

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/IElectricityMapsClient.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Client/IElectricityMapsClient.cs:19]: Post convo with Bill @JenMadiedo , will update to matchc better to EMaps

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Configuration/ElectricityMapsClientConfiguration.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Configuration/ElectricityMapsClientConfiguration.cs:39]: Post convo with Bill, will make it required because right now the SDK patsh all require authentication @JenMadiedo

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/HistoryCarbonIntensityData.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/src/Model/HistoryCarbonIntensityData.cs:70]: @JenMadiedo will remove default

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:7]: @bderusha owns

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 25]
```diff
@@ -0,0 +1,457 @@
...
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Moq;
+using System.Text.Json;
+
+namespace CarbonAware.DataSources.ElectricityMaps.Tests;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:17]: @bderusha  owns

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 26]
```diff
@@ -0,0 +1,457 @@
...
+
+    [Test]
+    public void AllPublicMethods_ThrowsWhenNoAuth()
+    {
+        // Arrange
+        CreateBasicClient(TestData.GetZonesAllowedJsonString(), "{}");
+
+        // Token Auth
+        var noAuthConfig_Token = new Mock<IOptionsMonitor<ElectricityMapsClientConfiguration>>();
+        noAuthConfig_Token.Setup(o => o.CurrentValue).Returns(() => new ElectricityMapsClientConfiguration());
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs:70]: @bderusha owns

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/CarbonAware.DataSources.ElectricityMaps.Tests.csproj (hunk 24 above)
[the remark this answers is not in the document]
Comment (pritipath) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/CarbonAware.DataSources.ElectricityMaps.Tests.csproj]: this is fixed

File: src/data/location-sources/custom-azure-zones.json (hunk 22 above)
Comment (juzuluag) [reply src/data/location-sources/custom-azure-zones.json]: @JenMadiedo , all resolved with this file.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs [hunk 27]
```diff
@@ -1,455 +0,0 @@
-using CarbonAware.Configuration;
```
Comment (bderusha) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs]: > **NOTE:**
> This file was renamed from `ElectricityMapClientTests.cs` (singular) to `ElectricityMapsClientTests.cs` (plural)
> Wish we got a better diff 😞

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs (hunk 2 above)
Comment (bderusha) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs]: Noted elsewhere for future cleanup.

**Review by bderusha — APPROVED** [review]
:shipit:

**Review by juzuluag — APPROVED** [review]
👍

**Comment by gfmatthews** [comment]
With those two review sign offs this now looks ready for review @vaughanknight

**Comment by bderusha** [comment]
> With those two review sign offs this now looks ready for review @vaughanknight

and @Willmish

**Review by vaughanknight — COMMENTED** [review]
Approved (apologies didn't click approve before)!
